p2p/listener.go

```diff
@@ -227,10 +231,55 @@ func (m *listener) providerStartConfigExchange(signerID identity.Identity, msg *
 		return fmt.Errorf("could not publish message via broker: %v", err)
 	}
 
-	m.setPendingConfig(publicIP, peerPubKey, privateKey, localPorts)
+	m.setPendingConfig(p2pConnectConfig{
+		publicIP:         publicIP,
+		localPorts:       localPorts,
+		privateKey:       privateKey,
+		peerPubKey:       peerPubKey,
+		upnpPortsRelease: portsRelease,
+		peerPublicIP:     "",
+		peerPorts:        nil,
+	})
 	return nil
 }
 
+func (m *listener) prepareLocalPorts(publicIP, outboundIP string) ([]int, []func(), error) {
+	// First acquire required only ports for needed n connections.
+	localPorts, err := acquireLocalPorts(m.portPool, requiredConnCount)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not acquire initial local ports: %v", err)
+	}
+	// Return these ports if provider is not behind NAT.
+	if outboundIP == publicIP {
+		return localPorts, nil, nil
+	}
+
+	// Try to add upnp ports mapping.
+	var portsRelease []func()
+	var portMappingOk bool
+	var portRelease func()
```

Comment:
@zolia I'm thinking what is the best way to handle cleanup for mapped ports. Maybe port mapper can hold mapped ports state globally and can be called on node shutdown with Close method. This way mapped ports can be handled in once place and we be sure they are closed. Also p2p channel holds the same release func so if session is destroyed gracefully(by triggering from consumer side) it will close p2p channel and release mapped ports. 

In generall I think such generic session cleanup on shutdown can be implemented in separate task https://github.com/mysteriumnetwork/node/issues/2032

Reply:
Global state basically IS needed, but not for shutdown cleanup. I think cleanup on session termination should do. Issue here is that destroys do not happen and nodes shuts down or are killed abruptly all the time. That will lead to upnp mapping poisoning on routers and eventually node will start to fail to allocate mappings.

My suggestion is to implement go routine which would poll actual state on a router and would sync it with a node's port mapping global state. This way we would be able to cleanup stale mappings. Question remains if we can make such polling reliably and if we can really check if mapping introduced previously is by our node or not. Maybe comment can hold some ids and be read from router.. need to check for similar stuff.